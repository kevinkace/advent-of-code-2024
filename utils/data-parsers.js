const EOL = "\n";
const DELIM = "   ";

export function splitToLists(data, { eol = EOL, delim = DELIM } = {}) {
    return data.split(eol).map(line => line.split(delim)).reduce((acc, cur) => {
        acc[0].push(Number(cur[0]));
        acc[1].push(Number(cur[1]));

        return acc;
    }, [[], []]);
}
