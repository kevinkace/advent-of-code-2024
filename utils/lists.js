export function getListDiffs([listA, listB]) {
    const diffs = [];

    for (let index = 0; index < listA.length; index++) {
        diffs.push(Math.abs(listA[index] - listB[index]));
    }

    return diffs;
}

export function sortLists(lists) {
    return lists.map(list => list.sort((a, b) => a > b ? 1 : -1));
}

export function sumList(list) {
    return list.reduce((acc, cur) => acc + cur, 0);
}

export function tally(list) {
    return list.reduce((acc, cur) => {
        const newVal = (acc.get(cur) || 0) + 1;

        acc.set(cur, newVal);

        return acc;
    }, new Map());
}

export function tallyProduct(tallyA, tallyB) {
    const products = [];

    const [shorterTally, longerTally] = [tallyA, tallyB].sort((tA, tB) => tA.length < tB.length ? 1 : -1);

    shorterTally.forEach((sTallyOccurrence, number) => {
        const lTallyOccurrence = longerTally.get(number) || 0;

        products.push(sTallyOccurrence * lTallyOccurrence * number);
    });

    return sumList(products);
}

export function arrToDiffs(arr) {
    return arr.reduce((acc, curr, index) => {
        const next = arr[index + 1];

        if (next === undefined) {
            return acc;
        }

        acc.push(next - curr);

        return acc;
    }, []);
}
