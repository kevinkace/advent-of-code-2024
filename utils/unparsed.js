
const EOL = "\r\n";

export function letterInbounds(string, { row, col }) {
    const rows = string.split(EOL);

    return row >= 0 && row < rows.length && col >= 0 && col < rows[0].length;
}

export function letterAt(string, { row, col }) {
    if (!letterInbounds(string, { row, col })) {
        return;
    }

    const rows = string.split(EOL);

    return rows[row][col];
}

export function numberRows(string) {
    return string.split(EOL)[0].length;
}
