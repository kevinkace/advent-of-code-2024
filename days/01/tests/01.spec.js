import { describe, expect, test } from "vitest";

import { splitToLists }                                          from "../../../utils/data-parsers";
import { getListDiffs, sortLists, sumList, tally, tallyProduct } from "../../../utils/lists";

import example1 from "../data/example-1";
import part1    from "../data/01-1";


describe("day 01-1", () => {
    test("example", () => {
        let [ listA, listB ] = splitToLists(example1);

        expect(listA).toEqual([ 3, 4, 2, 1, 3, 3 ]);

        [ listA, listB ] = sortLists([ listA, listB ]);


        expect(listA).toEqual([ 1, 2, 3, 3, 3, 4 ]);

        const diffs = getListDiffs([ listA, listB ]);

        expect(diffs).toEqual([ 2, 1, 0, 1, 2, 5 ]);

        expect(sumList(diffs)).toEqual(11);
    });

    test("part 1", () => {
        const lists = sortLists(splitToLists(part1));
        const diffs = getListDiffs(lists);
        const sum   = sumList(diffs);

        expect(sum).toEqual(2086478);
    });
});

describe("day 01-2", () => {
    test("tally", () => {
        const list       = [ 2, 4, 5, 3, 2, 3, 4, 4, 2 ];
        const [ sorted ] = sortLists([ list ]);

        expect(sorted).toMatchSnapshot();
        expect(tally(sorted)).toMatchSnapshot();
    });

    test("part 2 example", () => {
        const [ listA, listB ] = sortLists(splitToLists(example1));

        const tallyA = tally(listA);
        const tallyB = tally(listB);

        expect(tallyA).toMatchSnapshot();
        expect(tallyB).toMatchSnapshot();

        expect(tallyProduct(tallyA, tallyB)).toEqual(31);
    });

    test("part 2", () => {
        const [ tallyA, tallyB ] = sortLists(splitToLists(part1)).map(tally);

        expect(tallyProduct(tallyA, tallyB)).toEqual(24941624);
    });
});
