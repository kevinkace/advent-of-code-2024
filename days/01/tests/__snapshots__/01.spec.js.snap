// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 01-2 > part 2 example 1`] = `
Map {
  1 => 1,
  2 => 1,
  3 => 3,
  4 => 1,
}
`;

exports[`day 01-2 > part 2 example 2`] = `
Map {
  3 => 3,
  4 => 1,
  5 => 1,
  9 => 1,
}
`;

exports[`day 01-2 > tally 1`] = `
[
  2,
  2,
  2,
  3,
  3,
  4,
  4,
  4,
  5,
]
`;

exports[`day 01-2 > tally 2`] = `
Map {
  2 => 3,
  3 => 2,
  4 => 3,
  5 => 1,
}
`;
